File: src/python/demos/two_dim_tick_demo.py
from region import Region


def run_demo() -> None:
    height, width = 8, 8
    region = Region("two-d-tick-demo")

    input_index = region.add_input_layer_2d(height, width, gain=1.0, epsilon_fire=0.01)
    output_index = region.add_output_layer_2d(height, width, smoothing=0.0)

    unique_sources = region.connect_layers_windowed(
        input_index, output_index,
        kernel_h=3, kernel_w=3,
        stride_h=1, stride_w=1,
        padding="same",
    )
    print(f"unique_sources={unique_sources}")

    # Enable spatial metrics to compute active/centroid/bbox
    region.enable_spatial_metrics = True
    region.bind_input_2d("pixels", height, width, gain=1.0, epsilon_fire=0.01, attach_layers=[input_index])

    # Frame 1: single bright pixel near the center
    frame = [[0.0 for _ in range(width)] for _ in range(height)]
    frame[3][4] = 1.0
    metrics1 = region.tick_2d("pixels", frame)
    print(
        f"tick#1 delivered={metrics1.delivered_events} slots={metrics1.total_slots} synapses={metrics1.total_synapses} "
        f"active={metrics1.active_pixels} centroid=({metrics1.centroid_row:.3f},{metrics1.centroid_col:.3f}) bbox={metrics1.bbox}"
    )

    # Frame 2: move the bright pixel to exercise a different mapping
    frame[3][4] = 0.0
    frame[5][6] = 1.0
    metrics2 = region.tick_2d("pixels", frame)
    print(
        f"tick#2 delivered={metrics2.delivered_events} slots={metrics2.total_slots} synapses={metrics2.total_synapses} "
        f"active={metrics2.active_pixels} centroid=({metrics2.centroid_row:.3f},{metrics2.centroid_col:.3f}) bbox={metrics2.bbox}"
    )


if __name__ == "__main__":
    run_demo()

File: src/cpp/TwoDimTickDemo.cpp
#include "Region.h"
#include <iostream>
#include <vector>

using namespace grownet;

int main() {
    const int height = 8;
    const int width  = 8;

    Region region("two-d-tick-demo");

    int inputIndex  = region.addInputLayer2D(height, width, /*gain=*/1.0, /*epsilonFire=*/0.01);
    int outputIndex = region.addOutputLayer2D(height, width, /*smoothing=*/0.0);

    int uniqueSources = region.connectLayersWindowed(
        inputIndex, outputIndex,
        /*kernelH*/3, /*kernelW*/3,
        /*strideH*/1, /*strideW*/1,
        /*padding*/"same",
        /*feedback*/false);
    std::cout << "unique_sources=" << uniqueSources << "\n";

    // Bind input port to the 2D edge and wire to the pipeline
    region.bindInput2D("pixels", height, width, 1.0, 0.01, std::vector<int>{ inputIndex });

    // Build an 8x8 frame and light a single pixel
    std::vector<std::vector<double>> frame(height, std::vector<double>(width, 0.0));
    frame[3][4] = 1.0;

    auto m1 = region.tick2D("pixels", frame);
    std::cout << "tick#1 delivered=" << m1.delivered_events
              << " slots=" << m1.total_slots
              << " synapses=" << m1.total_synapses << "\n";

    frame[3][4] = 0.0;
    frame[5][6] = 1.0;
    auto m2 = region.tick2D("pixels", frame);
    std::cout << "tick#2 delivered=" << m2.delivered_events
              << " slots=" << m2.total_slots
              << " synapses=" << m2.total_synapses << "\n";

    return 0;
}

